util/auto_skip_compressor.cc

```diff
@@ -0,0 +1,118 @@
+      (static_cast<uint64_t>(opts.max_compressed_bytes_per_kb) *
+       uncompressed_block_data.size()) >>
+      10) {
+    rejected_count_++;
+  } else {
+    compressed_count_++;
+  }
+  attempted_compression_count_++;
+  if (attempted_compression_count_ >= kWindowSize) {
+    pred_rejection_percentage_ = static_cast<int>(
+        rejected_count_ * 100 / (compressed_count_ + rejected_count_));
+    attempted_compression_count_ = 0;
+    compressed_count_ = 0;
+    rejected_count_ = 0;
+  }
+  return true;
+}
+AutoSkipCompressorWrapper::AutoSkipCompressorWrapper(
+    std::unique_ptr<Compressor> compressor, const CompressionOptions& opts,
+    const CompressionType type)
+    : CompressorWrapper::CompressorWrapper(std::move(compressor)),
+      opts_(opts),
+      type_(type) {}
+
+Status AutoSkipCompressorWrapper::CompressBlock(
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, ManagedWorkingArea* wa) {
+  bool exploration =
+      Random::GetTLSInstance()->PercentTrue(kExplorationPercentage);
+  TEST_SYNC_POINT_CALLBACK(
+      "AutoSkipCompressorWrapper::CompressBlock::exploitOrExplore",
+      &exploration);
+  if (exploration) {
+    return CompressBlockAndRecord(uncompressed_data, compressed_output,
+                                  out_compression_type, wa);
+  } else {
+    auto predictor_ptr =
+        static_cast<AutoSkipCompressionContext*>(wa->get())->predictor_;
+    auto prediction = predictor_ptr->Predict();
+    if (prediction <= kProbabilityCutOff) {
+      // decide to compress
+      return CompressBlockAndRecord(uncompressed_data, compressed_output,
+                                    out_compression_type, wa);
+    } else {
+      // bypassed compression
+      *out_compression_type = kNoCompression;
+      return Status::OK();
+    }
+  }
+  return Status::OK();
+}
+
+Compressor::ManagedWorkingArea AutoSkipCompressorWrapper::ObtainWorkingArea() {
+  return ManagedWorkingArea(
+      static_cast<WorkingArea*>(new AutoSkipCompressionContext(type_, opts_)),
+      this);
+}
+void AutoSkipCompressorWrapper::ReleaseWorkingArea(WorkingArea* wa) {
+  delete static_cast<AutoSkipCompressionContext*>(wa);
+}
+
+Status AutoSkipCompressorWrapper::CompressBlockAndRecord(
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, ManagedWorkingArea* wa) {
+  Status status = wrapped_->CompressBlock(uncompressed_data, compressed_output,
+                                          out_compression_type, wa);
+  // determine if it was rejected or compressed
+  auto predictor_ptr =
+      static_cast<AutoSkipCompressionContext*>(wa->get())->predictor_;
+  predictor_ptr->Record(uncompressed_data, compressed_output, opts_);
+  return status;
+}
+
+const char* AutoSkipCompressorManager::Name() const {
+  // should have returned "AutoSkipCompressorManager" but we currently have an
+  // error so return return "AutoSkipCompressorManager";
```

Comment:
// error so return return "AutoSkipCompressorManager";

I believe you meant to say for now return wrapped compression manager name